### PR TITLE
Prevent double waitgroup decrement

### PR DESCRIPTION
**Pull request by zeripath**
A user in discourse has discovered that under heavy loads it is possible to cause a double decrement in the server connection waitgroup.

This PR prevents this.

### CODE REVIEW

**Review by guillep2k — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

File: modules/graceful/server.go [hunk 1]
```diff
@@ -241,12 +242,12 @@ func (wl *wrappedListener) File() (*os.File, error) {
 type wrappedConn struct {
 	net.Conn
 	server *Server
+	closed int32
```
Comment (guillep2k) [inline modules/graceful/server.go]:
```suggestion
	closed *int32
```

File: modules/graceful/server.go [hunk 2]
```diff
@@ -241,12 +242,12 @@ func (wl *wrappedListener) File() (*os.File, error) {
...
 }
 
 func (w wrappedConn) Close() error {
```
Comment (zeripath) [inline modules/graceful/server.go:251]:
```suggestion
func (w *wrappedConn) Close() error {
```

File: modules/graceful/server.go [hunk 3]
```diff
@@ -241,12 +242,12 @@ func (wl *wrappedListener) File() (*os.File, error) {
...
-	err := w.Conn.Close()
-	if err == nil {
+	if atomic.CompareAndSwapInt32(&w.closed, 0, 1) {
```
Comment (zeripath) [inline modules/graceful/server.go]:
```suggestion
	if atomic.CompareAndSwapInt32(&w.closed, 0, 1) {
```

File: modules/graceful/server.go [hunk 4]
```diff
@@ -241,12 +242,12 @@ func (wl *wrappedListener) File() (*os.File, error) {
 type wrappedConn struct {
 	net.Conn
 	server *Server
+	closed int32
 }
 
-func (w wrappedConn) Close() error {
-	err := w.Conn.Close()
-	if err == nil {
+func (w *wrappedConn) Close() error {
```
Comment (zeripath) [inline modules/graceful/server.go]:
```suggestion
func (w wrappedConn) Close() error {
```

File: modules/graceful/server.go [hunk 5]
```diff
@@ -241,12 +242,12 @@ func (wl *wrappedListener) File() (*os.File, error) {
...
+	if atomic.CompareAndSwapInt32(&w.closed, 0, 1) {
```
Comment (zeripath) [inline modules/graceful/server.go]:
```suggestion
	if atomic.CompareAndSwapInt32(w.closed, 0, 1) {
```

**Review by lafriks — APPROVED** [review]

**Review by guillep2k — APPROVED** [review]

**Comment by guillep2k** [comment]
The tests that failed before the latest commit (`*int32` vs. `int32`) show that this was a real problem.

**Pull request merged**